### ISSUE TITLE
Modernization

### ISSUE THREAD

**Issue by oliversalzburg** [colscott/lit-i18n#27]
Because the defined peer dependency range for this package is behind the currently released versions, it's holding back dependency upgrades. The dependency upgrades of this package itself seem to go unnoticed. The activity pattern of the maintainer of this package indicates very low activity.

Thus, I have forked this project, converted it to TypeScript, upgraded all dependencies, and published it under my own scope.

In case you want to merge any of the work, the branches have not diverged too far from another: https://github.com/oliversalzburg/lit-i18n

**Comment by colscott** [comment]
Hi @oliversalzburg, thank you for highlighting the peer dependency issue. This library is monitored and we do update the library. This library is used in an Enterprise environment and we update the dependencies as we need to, or when github issues are raised. If you or someone else can confirm that this library works with a newer version of i18next we can increase the peer dependency, or simply make a PR which we will gladly merge (this is open source after all).

We're not using TypeScript files in this repo. However, it is fully type safe using the TypeScript engine by using Type Script supported JavaScript and JSDoc. It saves a build step and makes debugging easier (no mapping files needed). You can still get nearly all of the Type Script features this way. Here are the supported [JSDoc tags](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).

**Comment by oliversalzburg** [comment]
The JSDoc tags that are currently applied through the source code do not match the actual types of i18next. Consuming their upstream types for type-checking also conveniently allows CI-based breaking change checking.

Purely widening the peer range is not sufficient to resolve the integration issues, as far as I could tell. Even using the latest i18next@21.x would result in the tests no longer working. I also had to disable the garbage collection, because the implementation of `isConnected()` seems to refer to nonexistent variables. Maybe this even requires more work in the near future.

**Comment by colscott** [comment]
@oliversalzburg, do you have the TS errors you get when consuming or a repo you were using that we can run to reproduce the issues?

We tried to remove the manual registry cleanup when upgrading to lit-html 2.x. But the manual registry cleanup was still required because we had cases where lit-html was not calling disconnect leading to memory leaks in our apps. Please let us know which variables were nonexistent and we'll be sure to fix those up.

**Comment by oliversalzburg** [comment]
This is what I was dealing with in `isConnected()`:

![image](https://github.com/colscott/lit-i18n/assets/1658949/cfa1ecf5-1771-4695-a771-fcd5ca2eb81e)

`part` is of type [`PartInfo`](https://lit.dev/docs/api/custom-directives/#PartInfo). This type has no member `element`, which breaks the entire function. From what I could tell, maybe a `Part` was expected here. But even that doesn't work, because `element` does not appear in every `Part`. I feel like there must have been a breaking change in i18next maybe.

I also saw that the `TranslateBase` constructor has the parameter marked as a `Part`, not a `PartInfo`. But when I looked at the current documentation for custom Lit directives, it seemed like accepting a `PartInfo` is actually correct.

I figured I would have to get back to this, maybe during the weekend, but maybe you'll be more successful :)

**Comment by colscott** [comment]
Managed to recreate the issue. Let me know if 4.0.4 works for you.
Bumped the major version as I made some breaking changes.

- lit-html is now a peer dependency
- implemented i18next module interface so that lit-i18n is initialized using i18next standard
- took the same approach as https://github.com/i18next/react-i18next/blob/master/package.json and left the peer dependency versions open

Use something like the following to bootstrap lit-i18n now:
```js
import i18next from 'i18next';
import { initLitI18n, translateWhen } from 'lit-i18n';

const initializePromise =  i18next
    .use(initLitI18n)
    .init({
        lng: 'en',
        resources: {
            en: {
                translation: {
                    hello: 'en-hello',
                },
            },
            fr: {
                translation: {
                    hello: 'fr-hello',
                },
            },
        },
    });

const t = (keys: string, options?: Record<string, unknown>) => translateWhen(initializePromise, keys, options);

const template = html`${t('hello')}`;
```

**Comment by oliversalzburg** [comment]
Nice! I'll check it out ASAP

**Comment by colscott** [comment]
Published 4.1.0. Given the way that lit-i18n is now initialized using the use(initLitI18n) approach, translateWhen is now redundant. Marked translateWhen as deprecated and updated the readme accordingly.

**Comment by oliversalzburg** [comment]
Looking great. Thanks for the quick work! I migrated back to this package, but couldn't test everything, because I need to resolve issues in other dependencies :(
